### ISSUE TITLE
Handle simple errors more gracefully

### ISSUE THREAD

**Issue by alexkreidler** [serde-rs/json#691]
TLDR: I got a completely useless and unhelpful error message from Serde when all that was wrong with the input was one comma. Please consider options to improve this.

## My issue
I had some json which roughly looked like this:
```
...json
    "requestor": {
        "plugin": "builtin",
        "config": {
            "timeout": "3",
            "headers": {
                "User-Agent": "bla",
            },
            "browser": [
                "Firefox",
                "random"
            ],
            "template": "default"
        }
    },
...
```
Notice the trailing comma after the `User-Agent` value.
I had this code, running in one of my tests:
```rust
let data = include_str!("../../../tests/basic.json");
let conf: Config = serde_json::from_str(data);
```
I thought it was some of my code, so I commented everything out except those two lines.

Then, the most descriptive error message I got was
```
thread 'agent::tests::configure' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', /home/alex/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/macros.rs:16:9
stack backtrace:
   0:     0x561bef204344 - backtrace::backtrace::libunwind::trace::h812748238d609e46
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   ...
test agent::tests::configure ... FAILED
```
Full stack trace here: https://pastemo.com/?bdcf7a23d3b6ffa3#8XR1PFew8yHpk2Bmytd8E7ZKTWgBVER72psVDC43MCtD.

After being dumbfounded for a while, because in the past Serde had told me exactly which fields were missing, etc, etc, I looked at the JSON file again.

After removing that one comma, the code worked perfectly.

## Proposal

Somehow, handle small errors like additional commas or malformed strings, etc in JSON. I understand this might decrease performance, so maybe having a separate implementation that does these sanity checks makes sense. 

Basically, do an initial validation of the JSON, or at least provide a reasonable error message overall, like: "cannot deserialize from string. please check your input"

**Comment by dtolnay** [comment]
I couldn't reproduce this; I get the error pointing out the trailing comma as I would expect.

```console
trailing comma at line 8 column 13
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f84d208a6eb8f8881d7640600c8d5f04

I'll close because this implies the crash you pasted is from elsewhere in your codebase, but feel free to share a compilable minimal repro if you are able to minimize to something in serde_json.

**Comment by Norlock** [comment]
How is this on line 8, column 13? I count line 7 and even if i put it on https://jsonformatter.org/ its line 7 column 37
